pkg/logutil/zap_text_encoder.go

```diff
@@ -0,0 +1,572 @@
+	}
+}
+
+func (enc *textEncoder) addKey(key string) {
+	enc.addElementSeparator()
+	enc.safeAddStringWithQuote(key)
+	enc.buf.AppendByte('=')
+}
+
+func (enc *textEncoder) addElementSeparator() {
+	last := enc.buf.Len() - 1
+	if last < 0 {
+		return
+	}
+	switch enc.buf.Bytes()[last] {
+	case '{', '[', ':', ',', ' ', '=':
+		return
+	default:
+		enc.buf.AppendByte(',')
+	}
+}
+
+func (enc *textEncoder) appendFloat(val float64, bitSize int) {
+	enc.addElementSeparator()
+	switch {
+	case math.IsNaN(val):
+		enc.buf.AppendString(`"NaN"`)
+	case math.IsInf(val, 1):
+		enc.buf.AppendString(`"+Inf"`)
+	case math.IsInf(val, -1):
+		enc.buf.AppendString(`"-Inf"`)
+	default:
+		enc.buf.AppendFloat(val, bitSize)
+	}
+}
+
+// safeAddString JSON-escapes a string and appends it to the internal buffer.
+// Unlike the standard library's encoder, it doesn't attempt to protect the
+// user from browser vulnerabilities or JSONP-related problems.
+func (enc *textEncoder) safeAddString(s string) {
+	for i := 0; i < len(s); {
+		if enc.tryAddRuneSelf(s[i]) {
+			i++
+			continue
+		}
+		r, size := utf8.DecodeRuneInString(s[i:])
+		if enc.tryAddRuneError(r, size) {
+			i++
+			continue
+		}
+		enc.buf.AppendString(s[i : i+size])
+		i += size
+	}
+}
+
+// safeAddStringWithQuote will automatically add quotoes.
+func (enc *textEncoder) safeAddStringWithQuote(s string) {
+	if !enc.needDoubleQuotes(s) {
+		enc.buf.AppendString(s)
+		return
+	}
+	enc.buf.AppendByte('"')
+	enc.safeAddString(s)
+	enc.buf.AppendByte('"')
+}
+
+// safeAddByteString is no-alloc equivalent of safeAddString(string(s)) for s []byte.
+func (enc *textEncoder) safeAddByteString(s []byte) {
+	for i := 0; i < len(s); {
+		if enc.tryAddRuneSelf(s[i]) {
+			i++
+			continue
+		}
+		r, size := utf8.DecodeRune(s[i:])
+		if enc.tryAddRuneError(r, size) {
+			i++
+			continue
+		}
+		enc.buf.Write(s[i : i+size])
+		i += size
+	}
+}
+
+// For https://github.com/tikv/rfcs/blob/master/text/2018-12-19-unified-log-format.md#log-fields-section.
+func (enc *textEncoder) needDoubleQuotes(s string) bool {
+	for i := 0; i < len(s); {
+		b := s[i]
+		if b <= 0x20 {
+			return true
+		}
+		switch b {
+		case '\\', '"', '[', ']', '=':
+			return true
+		case '\n', '\t', '\r':
+			return true
+		}
+		if b >= utf8.RuneSelf {
+			return true
+		}
+		i++
+	}
+	return false
+}
+
+// tryAddRuneSelf appends b if it is valid UTF-8 character represented in a single byte.
+func (enc *textEncoder) tryAddRuneSelf(b byte) bool {
+	if b >= utf8.RuneSelf {
+		return false
+	}
+	if 0x20 <= b && b != '\\' && b != '"' {
+		enc.buf.AppendByte(b)
+		return true
+	}
+	switch b {
+	case '\\', '"':
+		enc.buf.AppendByte('\\')
+		enc.buf.AppendByte(b)
+	case '\n':
+		enc.buf.AppendByte('\\')
+		enc.buf.AppendByte('n')
+	case '\r':
+		enc.buf.AppendByte('\\')
+		enc.buf.AppendByte('r')
+	case '\t':
+		enc.buf.AppendByte('\\')
+		enc.buf.AppendByte('t')
+
+	default:
+		// Encode bytes < 0x20, except for the escape sequences above.
```

Comment:
Looks like in JSON some of them are escaped as well, like `\f`, `\b`?

Reply:
The code here is the same as the standard library.
https://github.com/golang/go/blob/a2e79571a9d3dbe3cf10dcaeb1f9c01732219869/src/encoding/json/encode.go#L898-L916